File: src/test/java/io/lsdconsulting/junit5/extension/LsdExtensionTest.java
package io.lsdconsulting.junit5.extension;

import com.lsd.core.LsdContext;
import io.lsdconsulting.junit5.LsdExtension;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtensionContext;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.Optional;

import static com.lsd.core.properties.LsdProperties.OUTPUT_DIR;
import static com.lsd.core.properties.LsdProperties.get;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class LsdExtensionTest {
    private final ExtensionContext mockTestContext = mock(ExtensionContext.class);
    private final LsdContext lsdContext = LsdContext.getInstance();
    private final LsdExtension lsdExtension = new LsdExtension();
    private final File indexFile = new File(get(OUTPUT_DIR), "lsd-index.html");

    @BeforeEach
    public void setup() {
        lsdContext.clear();
    }

    @Test
    void approvalTestForTestFailedCallback() {
        configureMockContextToReturnDisplayNames();

        lsdExtension.testFailed(mockTestContext, new AssertionError("Expected <true> but was <false>"));

        Approvals.verify(lsdContext.completeReport("Failure Test Report").toFile());
    }

    @Test
    void approvalTestForTestSuccessfulCallback() throws IOException {
        configureMockContextToReturnDisplayNames();

        lsdExtension.testSuccessful(mockTestContext);

        Approvals.verify(copyOfFile(lsdContext.completeReport("Successful Test Report")));
    }

    @Test
    void approvalTestForTestAbortedCallback() throws IOException {
        configureMockContextToReturnDisplayNames();

        lsdExtension.testAborted(mockTestContext, new RuntimeException("Aborted for testing reasons"));

        Approvals.verify(copyOfFile(lsdContext.completeReport("Aborted Test Report")));
    }

    @Test
    void approvalTestForTestDisabledCallback() throws IOException {
        configureMockContextToReturnDisplayNames();

        lsdExtension.testDisabled(mockTestContext, Optional.of("Disabled reason"));

        Approvals.verify(copyOfFile(lsdContext.completeReport("Disabled Test Report")));
    }

    @Test
    void generateIndexAfterAllTestsInTestClass() {
        indexFile.delete();
        when(mockTestContext.getParent()).thenReturn(Optional.empty());
        when(mockTestContext.getDisplayName()).thenReturn("a display name");

        lsdExtension.afterAll(mockTestContext);

        assertThat(indexFile).exists();
    }

    @Test
    void doesNotGenerateIndexWhenTestIsNested() {
        indexFile.delete();
        when(mockTestContext.getParent()).thenReturn(Optional.of(mockTestContext));
        when(mockTestContext.getDisplayName()).thenReturn("a nested display name");

        lsdExtension.afterAll(mockTestContext);

        assertThat(indexFile).doesNotExist();
    }

    private void configureMockContextToReturnDisplayNames() {
        when(mockTestContext.getRequiredTestInstance()).thenReturn(this);
        when(mockTestContext.getDisplayName()).thenReturn("The test name");
        when(mockTestContext.getParent())
                .thenReturn(Optional.of(mockTestContext))
                .thenReturn(Optional.empty());
    }

    private File copyOfFile(Path original) throws IOException {
        Path copy = Path.of(original.getParent().toString(), "copy_" + original.getFileName());
        Files.copy(original, copy, StandardCopyOption.REPLACE_EXISTING);
        return copy.toFile();
    }
}

File: src/test/kotlin/io/lsdconsulting/junit5/extension/LsdExtensionTest.kt
package io.lsdconsulting.junit5.extension

import com.lsd.core.LsdContext
import com.lsd.core.properties.LsdProperties.OUTPUT_DIR
import com.lsd.core.properties.LsdProperties.get
import io.lsdconsulting.junit5.LsdExtension
import io.lsdconsulting.junit5.deCamelCase
import org.approvaltests.Approvals
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtensionContext
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import java.io.File
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.util.*

internal class LsdExtensionTest {
    private val mockTestContext = mock(ExtensionContext::class.java)
    private val lsdContext: LsdContext = LsdContext.instance
    private val lsdExtension = LsdExtension()
    private val indexFile = File(get(OUTPUT_DIR), "lsd-index.html")

    @BeforeEach
    fun setup() {
        lsdContext.clear()
    }

    @Test
    fun approvalTestForTestFailedCallback() {
        configureMockContextToReturnDisplayNames()
        lsdExtension.testFailed(mockTestContext, AssertionError("Expected <true> but was <false>"))
        Approvals.verify(lsdContext.completeReport("Failure Test Report").toFile())
    }

    @Test
    fun approvalTestForTestSuccessfulCallback() {
        configureMockContextToReturnDisplayNames()
        lsdExtension.testSuccessful(mockTestContext)
        Approvals.verify(copyOfFile(lsdContext.completeReport("Successful Test Report")))
    }

    @Test
    fun approvalTestForTestAbortedCallback() {
        configureMockContextToReturnDisplayNames()
        lsdExtension.testAborted(mockTestContext, RuntimeException("Aborted for testing reasons"))
        Approvals.verify(copyOfFile(lsdContext.completeReport("Aborted Test Report")))
    }

    @Test
    fun approvalTestForTestDisabledCallback() {
        configureMockContextToReturnDisplayNames()
        lsdExtension.testDisabled(mockTestContext, Optional.of("Disabled reason"))
        Approvals.verify(copyOfFile(lsdContext.completeReport("Disabled Test Report")))
    }

    @Test
    fun generateIndexAfterAllTestsInTestClass() {
        indexFile.delete()
        `when`(mockTestContext.parent).thenReturn(Optional.empty())
        `when`(mockTestContext.displayName).thenReturn("a display name")
        lsdExtension.afterAll(mockTestContext)
        assertThat(indexFile).exists()
    }

    @Test
    fun doesNotGenerateIndexWhenTestIsNested() {
        indexFile.delete()
        `when`(mockTestContext.parent).thenReturn(Optional.of(mockTestContext))
        `when`(mockTestContext.displayName).thenReturn("a nested display name")
        lsdExtension.afterAll(mockTestContext)
        assertThat(indexFile).doesNotExist()
    }

    @ParameterizedTest
    @CsvSource(
        "aCamelCaseExample,a camel case example",
        "CamelCase,camel case",
        "anITSolution,an it solution",
        "myClassUnderTestShouldDoSomethingNice(),my class under test should do something nice",
    )
    fun deCamelCase(input: String, expected: String) {
        assertThat(input.deCamelCase()).isEqualTo(expected)
    }

    private fun configureMockContextToReturnDisplayNames() {
        `when`(mockTestContext.requiredTestInstance).thenReturn(this)
        `when`(mockTestContext.displayName).thenReturn("aTestName()")
        `when`(mockTestContext.parent)
            .thenReturn(Optional.of(mockTestContext))
            .thenReturn(Optional.empty())
    }

    private fun copyOfFile(original: Path): File {
        val copy = Path.of(original.parent.toString(), "copy_" + original.fileName)
        Files.copy(original, copy, StandardCopyOption.REPLACE_EXISTING)
        return copy.toFile()
    }
}
